src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeployedToEnvironmentNotifierStep.java

```diff
@@ -52,53 +49,41 @@
 
     private final String environmentKey;
     private final String environmentName;
-    private final String environmentType;
-    private final String environmentUrl;
+    private final BitbucketDeploymentEnvironmentType environmentType;
+    private final URL environmentUrl;
```

Comment:
Using proper types rather than stringly typing things.

Reply:
I thought the goal was to ensure that we didn't end up with things being API (like `BitbucketDeploymentEnvironmentType`. That's why I changed them to strings in the first place

Reply:
I think I expressed myself unclearly, my apologies. 
The problem is that during deserialisation the constructor is never called. This means that any new field that is assigned in the constructor or has its value changed in some way in the constructor is an API breaking change. The `BitbucketDeploymentEnvironmentType` is an enum. The only way to break it is to remove an existing enum from the class.
I proposed against using our `BitbucketDeploymentEnvironment` as a field here because it is genuinely surprising for many people that the constructor is never run, and thus wouldn't consider changing the constructor or constructor assigned fields to be API breaking. That removing an enum from the `BitbucketDeploymentEnvironmentType` is API breaking should not be a surprise for anyone, therefore I think it is safe. I apologise that I didn't explain myself properly in my comment earlier.

Reply:
Ah right, makes sense.
> The only way to break it is to remove an existing enum from the class.

Other ways to break it are to rename it or move it. Using a string here avoids any potential problems

Reply:
This is true, I am a big fan of proper types if we can. It is always a trade-off between things. I can concede a `String` in this class, as long as all other classes take in the `BitbucketDeploymentEnvironmentType` and not a `String`.

Reply:
I hadn't thought about the breaking API angle when we talked about strongly typing the development environment (one can say the fields are unlikely to change but not enthusiastic to be proven wrong on that).

Reply:
yeah, same. I'm happy to keep as String in the Step, but have all other classes use proper types.